### ISSUE TITLE
Make gossip routines use TrySend, not Send

### ISSUE THREAD

**Issue by ValarDragon** [cometbft/cometbft#3186]
## Feature Request

Currently our gossip routines for vote and block gossip use `Send` not `TrySend`. `Send` is worse than `TrySend` in terms of our net CPU overhead due to instantiating multiple timers. The way select statements in golang work, is that every argument gets evaluated, and timer.NewTimer takes overhead to instantiate, and significant overhead to our scheduler to create. Its not clear to me we are ever benefitting in the gossip routine by using `Send` though. Lets think through the cases:

- Send does not block - Pretty clear to me that we wasted a timer allocation, and would have been better off doing a TrySend.
- Send blocks (channel buffer full, and we wait) - We would be better off not waiting, and retrying our loop on a more "refreshed" peer state, processing new inbound has votes. We will be more likely to get something helpful to the peer. Channel buffers are per peer x channel combo, so with modest capacity (which both votes and block parts have), we should not be hitting buffers. In the off chance we did, it should clear quickly.

### Problem Definition

Here is where we see it adding overhead to us:
![image](https://github.com/cometbft/cometbft/assets/6440154/a86d56e4-f61d-4cd9-a787-eaf3ebb60ae1)
(About 10% of the current gossip routine CPU time, but other concurrent work lowering overhead)

evidence of the go scheduler being high overhead right now. So likely we'd benefit from a few less timers.
![image](https://github.com/cometbft/cometbft/assets/6440154/6697aa94-004b-4997-a7c0-00189b2f4761)

### Proposal

Change gossip routines to use TrySend and not Send.

Is there any suggested QA process for this?

**Comment by cason** [comment]
Hey, I have in my backlog checking the usage of `Send` and `TrySend` in the consensus reactor. 

In particular because when `Send()` fails, the connection is broken (the timeout is like 10s).

**Comment by cason** [comment]
I have started investigating this behavior.

The `SendProposalSetHasProposal` sends in the blocking way a `Proposal` and possibly a `ProposalPOL` control message (only for re-proposed blocks). My opinion regarding this is:

- We can use `TrySend()` for the `Proposal`. If it fails, the method should return `false`, and we sleep for a bit. Currently, this method does not return anything and we continue (busy wait)
- Since we don't send `ProposalPOL` messages that often, I'd say to stick with `Send()` in this case. But only if the previous `TrySend()` has succeeded

**Comment by cason** [comment]
The `SendPartSetHasPart` method sends block parts. I would use `TrySend` here too and add a boolean return to the method. If `TrySend()` fails, we don't continue the main loop (i.e., we sleep for a bit). So, the same solution for the `Proposal` messages above.

**Comment by cason** [comment]
Finally, the `sendVoteSetHasVote` method send votes using the blocking version. It returns a boolean already, so we should go for `TrySend()` instead. We have the logic here to try again (continue, with another vote) if the method returns `true`, sleeping otherwise.

**Comment by cason** [comment]
> Send blocks (channel buffer full, and we wait) - We would be better off not waiting, and retrying our loop on a more "refreshed" peer state, processing new inbound has votes. 

Agree. Lets go for it, I don't think we are seeing relevant differences.

But I might need some testing to make sure that we try sending the same vote (one or a set) again if the first `TrySend` fails, just for sanity.

**Comment by cason** [comment]
> Send does not block - Pretty clear to me that we wasted a timer allocation, and would have been better off doing a TrySend.

I wonder whether `Send()` should have a timeout at all. We can replace the timeout by a `peer.Quit()` channel check, so that we leave the routine when the peer is dead.

**Comment by cason** [comment]
I think with the same solution we can address https://github.com/cometbft/cometbft/issues/2823 as well.
